File: CSharp/Eta/Applications/VAConsumer/SymbolListHandler.cs
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license      --
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
 *|                See the project's LICENSE.md for details.                  --
 *|           Copyright (C) 2022-2023 Refinitiv. All rights reserved.              --
 *|-----------------------------------------------------------------------------
 */

using LSEG.Eta.Common;
using LSEG.Eta.Codec;
using LSEG.Eta.Example.Common;
using LSEG.Eta.Rdm;
using LSEG.Eta.ValueAdd.Reactor;

namespace LSEG.Eta.ValueAdd.Consumer
{

    /// <summary>
    /// This is the symbol list handler for the ETA Value Add consumer application.
    /// </summary>
    ///
    /// It provides methods for sending the symbol list request(s) to a provider
    /// and processing the response(s). Method closing stream is also provided.
    ///
    internal class SymbolListHandler
    {
        public SymbolListRequest SymbolListRequest { get; set; } = new();
        public MarketPriceClose CloseMessage { get; set; } = new();
        public Codec.Buffer SymbolListName { get; set; } = new();
        public State State { get; set; } = new();
        public Qos Qos { get; set; } = new();
        public List<long> Capabilities { get; set; } = new List<long>();
        public int ServiceId { get; set; }
        public bool SnapshotRequest { get; set; }

        private const int SYMBOL_LIST_STREAM_ID_START = 400;
        private const int TRANSPORT_BUFFER_SIZE_REQUEST = 1000;
        private const int TRANSPORT_BUFFER_SIZE_CLOSE = 1000;

        private Map m_Map = new();
        private MapEntry m_MapEntry = new();
        private Codec.Buffer m_MapKey = new();
        private EncodeIterator m_EncodeIterator = new();
        private ReactorSubmitOptions m_SubmitOptions = new();


        public SymbolListHandler()
        {
        }

        private bool HasSymbolListCapability(List<long> capabilities)
        {
            return capabilities.Any((capability) => capability == (int)DomainType.SYMBOL_LIST);
        }

        public ReactorReturnCode SendRequest(ReactorChannel chnl, out ReactorErrorInfo? errorInfo)
        {
            /* check to see if the provider supports the symbol list domain */
            if (!HasSymbolListCapability(Capabilities))
            {
                errorInfo = new();
                errorInfo.Error.Text = "SYMBOL_LIST domain is not supported by the indicated provider";
                return ReactorReturnCode.FAILURE;
            }

            /* get a buffer for the item request */
            ITransportBuffer? msgBuf = chnl.GetBuffer(TRANSPORT_BUFFER_SIZE_REQUEST, false, out errorInfo);
            if (msgBuf == null)
            {
                return ReactorReturnCode.FAILURE;
            }

            /* initialize state management array */
            /* these will be updated as refresh and status messages are received */
            State.DataState(DataStates.NO_CHANGE);
            State.StreamState(StreamStates.UNSPECIFIED);

            /* encode symbol list request */
            SymbolListRequest.Clear();

            if (!SnapshotRequest)
                SymbolListRequest.Streaming = true;

            SymbolListRequest.SymbolListName.Data(SymbolListName.Data(), SymbolListName.Position, SymbolListName.Length);
            SymbolListRequest.StreamId = SYMBOL_LIST_STREAM_ID_START;
            SymbolListRequest.ServiceId = ServiceId;
            SymbolListRequest.HasServiceId = true;
            SymbolListRequest.Qos.IsDynamic = Qos.IsDynamic;
            SymbolListRequest.Qos.Rate(Qos.Rate());
            SymbolListRequest.Qos.Timeliness(Qos.Timeliness());
            SymbolListRequest.HasQos = true;
            SymbolListRequest.PriorityClass = 1;
            SymbolListRequest.PriorityCount = 1;

            m_EncodeIterator.Clear();
            m_EncodeIterator.SetBufferAndRWFVersion(msgBuf, chnl.MajorVersion, chnl.MinorVersion);

            Console.WriteLine(SymbolListRequest.ToString());
            CodecReturnCode ret = SymbolListRequest.Encode(m_EncodeIterator);
            if (ret != CodecReturnCode.SUCCESS)
            {
                return ReactorReturnCode.FAILURE;
            }

            return chnl.Submit(msgBuf, m_SubmitOptions, out errorInfo);
        }

        public ReactorReturnCode ProcessResponse(Msg msg, DecodeIterator dIter, DataDictionary dictionary, out ReactorErrorInfo? errorInfo)
        {
            m_Map.Clear();
            m_MapEntry.Clear();
            m_MapKey.Clear();
            switch (msg.MsgClass)
            {
                case MsgClasses.REFRESH:
                    IRefreshMsg refreshMsg = (IRefreshMsg)msg;
                    State.DataState(refreshMsg.State.DataState());
                    State.StreamState(refreshMsg.State.StreamState());
                    return HandleUpdate(msg, dIter, m_Map, m_MapEntry, m_MapKey, out errorInfo);

                case MsgClasses.UPDATE:
                    return HandleUpdate(msg, dIter, m_Map, m_MapEntry, m_MapKey, out errorInfo);

                case MsgClasses.STATUS:
                    IStatusMsg statusMsg = (IStatusMsg)msg;
                    Console.WriteLine("Received Item StatusMsg for stream " + msg.StreamId);
                    if (statusMsg.CheckHasState())
                    {
                        State.DataState(statusMsg.State.DataState());
                        State.StreamState(statusMsg.State.StreamState());
                        State.Text(statusMsg.State.Text());
                        Console.WriteLine("    " + State);
                    }
                    errorInfo = null;
                    return ReactorReturnCode.SUCCESS;
                case MsgClasses.ACK:
                    HandleAck(msg);
                    errorInfo = null;
                    return ReactorReturnCode.SUCCESS;
                default:
                    Console.WriteLine("Received Unhandled Item Msg Class: " + msg.MsgClass);
                    break;
            }

            errorInfo = null;
            return ReactorReturnCode.SUCCESS;
        }

        protected ReactorReturnCode HandleUpdate(Msg msg, DecodeIterator dIter, Map map,
                MapEntry mapEntry, Codec.Buffer mapKey, out ReactorErrorInfo? errorInfo)
        {
            // print the name of the symbolist and the domain
            Console.WriteLine(SymbolListName.ToString() +
                    "\nDOMAIN: " + DomainTypes.ToString(msg.DomainType));

            CodecReturnCode ret = map.Decode(dIter);
            if (ret != CodecReturnCode.SUCCESS)
            {
                errorInfo = new();
                errorInfo.Error.Text = "SymbolListHandler failed to decode update";
                Console.WriteLine($"DecodeMap() failed: <{ret}>");
                return ReactorReturnCode.FAILURE;
            }

            // decode the map
            while ((ret = mapEntry.Decode(dIter, mapKey)) != CodecReturnCode.END_OF_CONTAINER)
            {
                if (ret != CodecReturnCode.SUCCESS)
                {
                    errorInfo = new();
                    errorInfo.Error.Text = "SymbolListHandler failed to decode update";
                    Console.WriteLine($"DecodeMapEntry() failed: <{ret}>");
                    return ReactorReturnCode.FAILURE;
                }
                Console.WriteLine(mapKey.ToString() + "\t" +
                            MapEntryActionToString(mapEntry.Action));
            }

            errorInfo = null;
            return ReactorReturnCode.SUCCESS;
        }

        protected ReactorReturnCode HandleAck(Msg msg)
        {
            /*
             * although this application only posts on Market Price, ACK handler is
             * provided for other domains to allow user to extend and post on MBO
             * and MBP domains
             */
            Console.WriteLine("Received AckMsg for stream " + msg.StreamId);
            IMsgKey key = msg.MsgKey;

            //print out item name from key if it has it
            if (key != null && key.CheckHasName())
            {
                Console.WriteLine(key.Name + "\nDOMAIN: " +
                            DomainTypes.ToString(msg.DomainType));
            }

            IAckMsg ackMsg = (IAckMsg)msg;
            Console.WriteLine("\tackId=" + ackMsg.AckId
                + (ackMsg.CheckHasSeqNum() ? "\tseqNum=" + ackMsg.SeqNum : "")
                + (ackMsg.CheckHasNakCode() ? "\tnakCode=" + ackMsg.NakCode : "")
                + (ackMsg.CheckHasText() ? "\ttext=" + ackMsg.Text.ToString() : ""));
            return ReactorReturnCode.SUCCESS;
        }


        /// <summary>
        /// Close the symbol list stream.
        /// </summary>
        public ReactorReturnCode CloseStream(ReactorChannel chnl, out ReactorErrorInfo? errorInfo)
        {
            /*
             * we only want to close a stream if it was not already closed (e.g.
             * rejected by provider, closed via refresh or status, or redirected)
             */
            if (State.StreamState() != StreamStates.OPEN
                && State.DataState() != DataStates.OK)
            {
                errorInfo = null;
                return ReactorReturnCode.SUCCESS;
            }

            //get a buffer for the item close
            ITransportBuffer? msgBuf = chnl.GetBuffer(TRANSPORT_BUFFER_SIZE_CLOSE, false, out errorInfo);

            if (msgBuf == null)
            {
                return ReactorReturnCode.FAILURE;
            }

            //encode item close
            CloseMessage.Clear();

            CloseMessage.StreamId = SYMBOL_LIST_STREAM_ID_START;
            CloseMessage.DomainType = (int)DomainType.SYMBOL_LIST;
            m_EncodeIterator.Clear();
            m_EncodeIterator.SetBufferAndRWFVersion(msgBuf, chnl.MajorVersion, chnl.MinorVersion);

            CodecReturnCode ret = CloseMessage.Encode(m_EncodeIterator);
            if (ret != CodecReturnCode.SUCCESS)
            {
                if(errorInfo is null)
                {
                    errorInfo = new ReactorErrorInfo();
                }
                errorInfo.Error.Text = $"EncodeSymbolListClose(): Failed <{ret}>";
                return ReactorReturnCode.FAILURE;
            }
            return chnl.Submit(msgBuf, m_SubmitOptions, out errorInfo);
        }


        private string MapEntryActionToString(MapEntryActions mapEntryAction)
        {
            switch (mapEntryAction)
            {
                case MapEntryActions.UPDATE:
                    return "UPDATE";
                case MapEntryActions.ADD:
                    return "ADD";
                case MapEntryActions.DELETE:
                    return "DELETE";
                default:
                    return "Unknown Map Entry Action";
            }
        }
    }
}

File: Java/Eta/Applications/Examples/src/main/java/com/refinitiv/eta/valueadd/examples/consumer/SymbolListHandler.java
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.
 *|                See the project's LICENSE.md for details.
 *|           Copyright (C) 2020,2022,2024 LSEG. All rights reserved.
 *|-----------------------------------------------------------------------------
 */

package com.refinitiv.eta.valueadd.examples.consumer;

import java.util.ArrayList;
import java.util.List;

import com.refinitiv.eta.codec.AckMsg;
import com.refinitiv.eta.codec.Buffer;
import com.refinitiv.eta.codec.CodecFactory;
import com.refinitiv.eta.codec.CodecReturnCodes;
import com.refinitiv.eta.codec.DataDictionary;
import com.refinitiv.eta.codec.DataStates;
import com.refinitiv.eta.codec.DecodeIterator;
import com.refinitiv.eta.codec.EncodeIterator;
import com.refinitiv.eta.codec.Map;
import com.refinitiv.eta.codec.MapEntry;
import com.refinitiv.eta.codec.MapEntryActions;
import com.refinitiv.eta.codec.Msg;
import com.refinitiv.eta.codec.MsgClasses;
import com.refinitiv.eta.codec.MsgKey;
import com.refinitiv.eta.codec.Qos;
import com.refinitiv.eta.codec.RefreshMsg;
import com.refinitiv.eta.codec.State;
import com.refinitiv.eta.codec.StatusMsg;
import com.refinitiv.eta.codec.StreamStates;
import com.refinitiv.eta.shared.rdm.marketprice.MarketPriceClose;
import com.refinitiv.eta.shared.rdm.symbollist.SymbolListRequest;
import com.refinitiv.eta.rdm.DomainTypes;
import com.refinitiv.eta.transport.TransportBuffer;
import com.refinitiv.eta.valueadd.reactor.ReactorChannel;
import com.refinitiv.eta.valueadd.reactor.ReactorErrorInfo;
import com.refinitiv.eta.valueadd.reactor.ReactorFactory;
import com.refinitiv.eta.valueadd.reactor.ReactorSubmitOptions;

/*
 * This is the symbol list handler for the ETA Value Add consumer application.
 * It provides methods for sending the symbol list request(s) to a provider
 * and processing the response(s). Method closing stream is also provided.
 */
class SymbolListHandler
{
    private final int SYMBOL_LIST_STREAM_ID_START = 400;
    private final int TRANSPORT_BUFFER_SIZE_REQUEST = 1000;
    private final int TRANSPORT_BUFFER_SIZE_CLOSE = 1000;

    private SymbolListRequest symbolListRequest;
    private MarketPriceClose closeMessage;
    private Buffer symbolListName;
    private State state;
    private Qos qos;
    private List<Long> capabilities;
    private int serviceId;
    private Map map = CodecFactory.createMap();
    private MapEntry mapEntry = CodecFactory.createMapEntry();
    private Buffer mapKey = CodecFactory.createBuffer();
    private EncodeIterator encIter = CodecFactory.createEncodeIterator();

    private boolean snapshotRequested;

    private ReactorSubmitOptions submitOptions = ReactorFactory.createReactorSubmitOptions();

    SymbolListHandler()
    {
        state = CodecFactory.createState();
        qos = CodecFactory.createQos();
        capabilities = new ArrayList<Long>();
        symbolListName = CodecFactory.createBuffer();
        symbolListRequest = new SymbolListRequest();
        closeMessage = new MarketPriceClose();
    }

    void snapshotRequest(boolean snapshotRequested)
    {
        this.snapshotRequested = snapshotRequested;
    }

    Buffer symbolListName()
    {
        return symbolListName;
    }

    int serviceId()
    {
        return serviceId;
    }

    void serviceId(int serviceId)
    {
        this.serviceId = serviceId;
    }

    Qos qos()
    {
        return qos;
    }

    List<Long> capabilities()
    {
        return capabilities;
    }

    private boolean hasSymbolListCapability(List<Long> capabilities)
    {
        for (Long capability : capabilities)
        {
            if (capability == DomainTypes.SYMBOL_LIST)
                return true;
        }
        return false;
    }

    int sendRequest(ReactorChannel chnl, ReactorErrorInfo errorInfo)
    {
        /* check to see if the provider supports the symbol list domain */
        if (!hasSymbolListCapability(capabilities()))
        {
            errorInfo.error().text("SYMBOL_LIST domain is not supported by the indicated provider");
            return CodecReturnCodes.FAILURE;
        }

        /* get a buffer for the item request */
        TransportBuffer msgBuf = chnl.getBuffer(TRANSPORT_BUFFER_SIZE_REQUEST, false, errorInfo);
        if (msgBuf == null)
        {
            return CodecReturnCodes.FAILURE;
        }

        /* initialize state management array */
        /* these will be updated as refresh and status messages are received */
        state.dataState(DataStates.NO_CHANGE);
        state.streamState(StreamStates.UNSPECIFIED);

        /* encode symbol list request */
        symbolListRequest.clear();

        if (!snapshotRequested)
            symbolListRequest.applyStreaming();
        symbolListRequest.symbolListName().data(symbolListName.data(), symbolListName.position(), symbolListName.length());
        symbolListRequest.streamId(SYMBOL_LIST_STREAM_ID_START);
        symbolListRequest.serviceId(serviceId());
        symbolListRequest.applyHasServiceId();
        symbolListRequest.qos().dynamic(qos.isDynamic());
        symbolListRequest.qos().rate(qos.rate());
        symbolListRequest.qos().timeliness(qos.timeliness());
        symbolListRequest.applyHasQos();
        symbolListRequest.priority(1, 1);

        encIter.clear();
        encIter.setBufferAndRWFVersion(msgBuf, chnl.majorVersion(), chnl.minorVersion());

        System.out.println(symbolListRequest.toString());
        int ret = symbolListRequest.encode(encIter);
        if (ret != CodecReturnCodes.SUCCESS)
        {
            return ret;
        }

        return chnl.submit(msgBuf, submitOptions, errorInfo);
    }

    @SuppressWarnings("fallthrough")
	int processResponse(Msg msg, DecodeIterator dIter, DataDictionary dictionary)
    {
        map.clear();
        mapEntry.clear();
        mapKey.clear();
        switch (msg.msgClass())
        {
            case MsgClasses.REFRESH:
                RefreshMsg refreshMsg = (RefreshMsg)msg;
                state.dataState(refreshMsg.state().dataState());
                state.streamState(refreshMsg.state().streamState());
            case MsgClasses.UPDATE:
                return handleUpdate(msg, dIter, map, mapEntry, mapKey);

            case MsgClasses.STATUS:
                StatusMsg statusMsg = (StatusMsg)msg;
                System.out.println("Received Item StatusMsg for stream " + msg.streamId());
                if (statusMsg.checkHasState())
                {
                    this.state.dataState(statusMsg.state().dataState());
                    this.state.streamState(statusMsg.state().streamState());
                    System.out.println("    " + state);
                }
                return CodecReturnCodes.SUCCESS;
            case MsgClasses.ACK:
                handleAck(msg);
                return CodecReturnCodes.SUCCESS;
            default:
                System.out.println("Received Unhandled Item Msg Class: " + msg.msgClass());
                break;
        }

        return CodecReturnCodes.SUCCESS;
    }

    protected int handleUpdate(Msg msg, DecodeIterator dIter, Map map,
            MapEntry mapEntry, Buffer mapKey)
    {
        // print the name of the symbolist and the domain
        System.out.println(symbolListName.toString() +
                "\nDOMAIN: " + DomainTypes.toString(msg.domainType()));

        int ret = map.decode(dIter);
        if (ret != CodecReturnCodes.SUCCESS)
        {
            System.out.println("DecodeMap() failed: < " + CodecReturnCodes.toString(ret) + ">");
            return ret;
        }

        // decode the map
        while ((ret = mapEntry.decode(dIter, mapKey)) != CodecReturnCodes.END_OF_CONTAINER)
        {
            if (ret != CodecReturnCodes.SUCCESS)
            {
                System.out.println("DecodeMapEntry() failed: < " + CodecReturnCodes.toString(ret) + ">");
                return ret;
            }
            System.out.println(mapKey.toString() + "\t" +
                        mapEntryActionToString(mapEntry.action()));
        }

        return CodecReturnCodes.SUCCESS;
    }

    protected int handleAck(Msg msg)
    {
        /*
         * although this application only posts on Market Price, ACK handler is
         * provided for other domains to allow user to extend and post on MBO
         * and MBP domains
         */
        System.out.println("Received AckMsg for stream " + msg.streamId());
        MsgKey key = msg.msgKey();

        //print out item name from key if it has it
        if (key != null && key.checkHasName())
        {
            System.out.println(key.name() + "\nDOMAIN: " +
                        DomainTypes.toString(msg.domainType()));
        }

        AckMsg ackMsg = (AckMsg)msg;
        System.out.println("\tackId=" + ackMsg.ackId() + (ackMsg.checkHasSeqNum() ? "\tseqNum=" + ackMsg.seqNum() : "") + (ackMsg.checkHasNakCode() ? "\tnakCode=" + ackMsg.nakCode() : "") + (ackMsg.checkHasText() ? "\ttext=" + ackMsg.text().toString() : ""));
        return CodecReturnCodes.SUCCESS;
    }

    /*
     * Close the symbol list stream.
     */
    int closeStream(ReactorChannel chnl, ReactorErrorInfo errorInfo)
    {
        /*
         * we only want to close a stream if it was not already closed (e.g.
         * rejected by provider, closed via refresh or status, or redirected)
         */
        if (state.streamState() != StreamStates.OPEN && state.dataState() != DataStates.OK)
            return CodecReturnCodes.SUCCESS;

        //get a buffer for the item close
        TransportBuffer msgBuf = chnl.getBuffer(TRANSPORT_BUFFER_SIZE_CLOSE, false, errorInfo);

        if (msgBuf == null)
        {
            return CodecReturnCodes.FAILURE;
        }

        //encode item close
        closeMessage.clear();

        closeMessage.streamId(SYMBOL_LIST_STREAM_ID_START);
        closeMessage.domainType(DomainTypes.SYMBOL_LIST);
        encIter.clear();
        encIter.setBufferAndRWFVersion(msgBuf, chnl.majorVersion(), chnl.minorVersion());

        int ret = closeMessage.encode(encIter);
        if (ret != CodecReturnCodes.SUCCESS)
        {
            errorInfo.error().text("encodeSymbolListClose(): Failed <" + CodecReturnCodes.toString(ret) + ">");
            return ret;
        }
        return chnl.submit(msgBuf, submitOptions, errorInfo);
    }

    private String mapEntryActionToString(int mapEntryAction)
    {
        switch (mapEntryAction)
        {
            case MapEntryActions.UPDATE:
                return "UPDATE";
            case MapEntryActions.ADD:
                return "ADD";
            case MapEntryActions.DELETE:
                return "DELETE";
            default:
                return "Unknown Map Entry Action";
        }
    }
}
